modules/data-streams/src/main/java/org/elasticsearch/datastreams/action/TransportReindexDataStreamIndexAction.java

```diff
@@ -0,0 +1,214 @@
+import java.util.Set;
+
+public class TransportReindexDataStreamIndexAction extends HandledTransportAction<
+    ReindexDataStreamIndexAction.Request,
+    ReindexDataStreamIndexAction.Response> {
+
+    private static final Logger logger = LogManager.getLogger(TransportReindexDataStreamIndexAction.class);
+
+    // copied from
+    // https://github.com/elastic/kibana/blob/8a8363f02cc990732eb9cbb60cd388643a336bed/x-pack/plugins/upgrade_assistant/server/lib/reindexing/index_settings.ts#L155
+    private static final Set<String> DISALLOWED_SETTINGS = Set.of(
+        // Private ES settings
+        "index.allocation.existing_shards_allocator",
+        "index.blocks.write",
+        "index.creation_date",
+        "index.frozen",
+        "index.history.uuid",
+        "index.merge.enabled",
+        "index.provided_name",
+        "index.resize.source.name",
+        "index.resize.source.uuid",
+        "index.routing.allocation.initial_recovery._id",
+        "index.search.throttled",
+        "index.source_only",
+        "index.shrink.source.name",
+        "index.shrink.source.uuid",
+        "index.uuid",
+        "index.verified_before_close",
+        // TODO most snapshot settings will need to be set to a different value in the result index
+        "index.store.snapshot.repository_name",
+        "index.store.snapshot.snapshot_name",
+        "index.store.snapshot.snapshot_uuid",
+        "index.store.snapshot.index_name",
+        "index.store.snapshot.index_uuid",
+
+        // TODO make sure has new value
+        "index.version.created",
+
+        // Deprecated in 9.0
+        "index.version.upgraded"
+    );
+
+    private static final IndicesOptions IGNORE_MISSING_OPTIONS = IndicesOptions.fromOptions(true, true, false, false);
+    private final ClusterService clusterService;
+    private final Client client;
+
+    @Inject
+    public TransportReindexDataStreamIndexAction(
+        TransportService transportService,
+        ClusterService clusterService,
+        ActionFilters actionFilters,
+        Client client
+    ) {
+        super(
+            ReindexDataStreamIndexAction.NAME,
+            false,
+            transportService,
+            actionFilters,
+            ReindexDataStreamIndexAction.Request::new,
+            transportService.getThreadPool().executor(ThreadPool.Names.GENERIC)
+        );
+        this.clusterService = clusterService;
+        this.client = client;
+    }
+
+    @Override
+    protected void doExecute(
+        Task task,
+        ReindexDataStreamIndexAction.Request request,
+        ActionListener<ReindexDataStreamIndexAction.Response> listener
+    ) {
+        var sourceIndexName = request.getSourceIndex();
+        var destIndexName = generateDestIndexName(sourceIndexName);
+        IndexMetadata sourceIndex = clusterService.state().getMetadata().index(sourceIndexName);
+
+        if (sourceIndex.getCreationVersion().isLegacyIndexVersion() == false) {
+            logger.warn(
```

Comment:
We don't want this warning if we use this in the future for the non-upgrade case.

Reply:
Although I guess this transport action is specific to the upgrade case.

Reply:
I figure I'll leave this warning in for now and can remove if we extend to more general reindexing.